pkg/driver/driver_azure.go

```diff
@@ -1035,20 +1047,64 @@ func (clients *azureDriverClients) checkOrphanDisks(ctx context.Context, resourc
 }
 
 func (clients *azureDriverClients) deleteNIC(ctx context.Context, resourceGroupName string, nicName string) error {
+	var (
+		nicDeletionTimeout     = 10 * time.Minute
+		nicDeletionCtx, cancel = context.WithTimeout(ctx, nicDeletionTimeout)
+	)
+	defer cancel()
+
 	klog.V(2).Infof("NIC delete started for %q", nicName)
 	defer klog.V(2).Infof("NIC deleted for %q", nicName)
 
-	future, err := clients.nic.Delete(ctx, resourceGroupName, nicName)
+	future, err := clients.nic.Delete(nicDeletionCtx, resourceGroupName, nicName)
 	if err != nil {
 		return onARMAPIErrorFail(prometheusServiceNIC, err, "nic.Delete")
 	}
-	if err := future.WaitForCompletionRef(ctx, clients.nic.Client); err != nil {
+	if err := future.WaitForCompletionRef(nicDeletionCtx, clients.nic.Client); err != nil {
+		return onARMAPIErrorFail(prometheusServiceNIC, err, "nic.Delete")
+	}
+
+	if err := clients.waitUntilNICIsDeleted(nicDeletionCtx, resourceGroupName, nicName); err != nil {
 		return onARMAPIErrorFail(prometheusServiceNIC, err, "nic.Delete")
 	}
+
 	onARMAPISuccess(prometheusServiceNIC, "NIC deletion was successful for %s", nicName)
 	return nil
 }
 
+func (clients *azureDriverClients) waitUntilNICIsDeleted(ctx context.Context, resourceGroupName string, nicName string) error {
```

Comment:
It might be a good idea to split this function into two.

1. `waitUntil(ctx context.Context, conditionFn func() bool)` which implements just implements the loop logic (with the backoff) and just calls the `conditionFn` function inside the loop in every iteration as long as `conditionFn` return `false` and quits the loop if the `conditionFn` returns `true`.
2. `conditionNICDeleted(ctx context.Context, resourceGroupName string, nicName string) func() bool` which returns a function which implements the get call and the check if it is not found. Something like...
```go
func (clients *azureDriverClients) conditionNICDeleted(ctx context.Context, resourceGroupName string, nicName string) func() bool {
	return func() bool {
		_, err := clients.nic.Get(ctx, resourceGroupName, nicName, "")

		if err != nil && isResourceNotFoundError(err) {
			// NIC doesn't exist, hence deletion is successful
			return true
		}
		
		return true
	}
}
```

This way, you can reuse the code easily if we need to use similar logic elsewhere later (creation or ever for disks etc.). WDYT?

Reply:
Agreed @amshuman-kr . Sure shall make this change.

Reply:
Some of the backoff library already do something like this. Then you have to just write the condition part. Not sure of the client-go.Backoff has something like this or not.

Reply:
I have made `waitUntil()` as a helper package method to be able to be reused in the future [here](https://github.com/gardener/machine-controller-manager/pull/594/files#diff-9dbc206e5c2d8e9011611b1f513e2d2e73140382dae88afe6969374aab75a9edR29-R56). Also [checkNICStatus()](https://github.com/gardener/machine-controller-manager/pull/594/files#diff-36365e28193d28c39f36ac4d9172fbf8eac69553361f6179fa3184f9cad9eb0cR1050-R1075) to get the NIC status while being deleted (also can be used for creation).